File: 1209-remove-all-adjacent-duplicates-in-string-ii/1209-remove-all-adjacent-duplicates-in-string-ii.cpp
class Solution {
public:
    string removeDuplicates(string s, int k) {
        stack<pair<char,int>>st;
        
        for(auto &ch:s){
          if(!st.empty() and st.top().first==ch)
              st.top().second+=1;
          else
               st.push({ch,1});
          if(st.top().second==k)
               st.pop();
        }
        string res;
        while(!st.empty()){
            int cnt=st.top().second;
            while(cnt--) res+=st.top().first;
            st.pop();
        }
        reverse(res.begin(),res.end());
        return res;
    }
};

File: 1209-remove-all-adjacent-duplicates-in-string-ii/1209-remove-all-adjacent-duplicates-in-string-ii.js
/**
 * @param {string} s
 * @param {number} k
 * @return {string}
 */
var removeDuplicates = function (s, k) {
    const stack = []; // {char, cnt};

    for (let char of s) {
        if (stack.length && stack.at(-1).char === char) {
            let prev = stack.at(-1);
            prev.cnt += 1;
            if (prev.cnt === k) {
                stack.pop();
            }
        }
        else {
            stack.push({ char, cnt: 1 });
        }
    }

    return stack.map(item => item.char.repeat(item.cnt)).join('');
};